Document V&V, automated V&V, and fuzzy checking (#1181)
ihmeuw/vivarium_research: docs/source/model_design/vivarium_features/automated_v_and_v/index.rst

and generate an unambiguous verdict on whether or not there is an issue.

.. note:: 
  In the future, we could explore having an in-between result, where the automated process flags
  a *potential* issue for human review.
  For now, we stick to a binary result.

There are some process questions to work out about this transformation:

* Who investigates the cause of automated V&V failures -- research or engineering? Does it depend on the type of failure?
* How frequently does automated V&V need to be run, and with what population size/number of draws/etc?
    * Should we consider separate "tiers" of automated checks that require different amounts of computational resources?
* What sort of sensitivity/specificity tradeoff do we want for our automated V&V?
  This likely depends on both of the previous questions.

PRL case study: integration tests
+++++++++++++++++++++++++++++++++

We have automated some V&V checks on the PRL project by adding integration tests
to `the simulation repository <https://github.com/ihmeuw/vivarium_census_prl_synth_pop/>`_.
These tests run the simulation for a given number of time steps and check certain conditions

> with explicit Python :code:`assert` statements, which fail in unexpected situations.

Reviewer: ```suggestion
with explicit Python :code:`assert` statements, which fail when the model exhibits unexpected behavior.
```
Changing this because before it could be interpreted that we cannot predict when our tests will fail :)

For example, there is a test that checks that each non-GQ household in the simulation has exactly one

> living reference person.

Reviewer: ```suggestion
living reference person, a condition which should always be true.
```

> The integration tests have access to intermediate simulation state, not only simulation outputs, and can check

Reviewer: ```suggestion
The integration tests have access to intermediate simulation states, not only simulation outputs, and can check
```

In the future, this could be done by continuous integration (CI) and required before merging; at present,
the simulation artifact is only available internally to IHME, which makes using it from an external
CI server impossible.

Tests have been developed by the engineering team to cover most of the simulation's functionality,
though test coverage is not complete.
Notably, none of the observers currently have integration tests.

Fuzzy checking
--------------

It is fairly straightforward to automate checks such as
"a household *always* has exactly one living reference person."
It is more difficult to check that a value is correct when the value in the
simulation is subject to stochastic variation.
We clearly cannot check that the rate of events in the simulation is **exactly** the expected rate,
or that a mean value of a continuous risk factor is **exactly** as expected.
Nor can we use an approximation like "within X% of the true value,"
since how much a value deviates due to chance depends on our population size.
The difficulty of this problem is part of why, in the manual V&V process, we usually check such values visually.

> By building a proof of concept for the PRL simulation, we have begun to formalize this to enable

Reviewer 2: ```suggestion
By building a proof-of-concept for the PRL simulation, we have begun to formalize this to enable
```
I don't want you to think i am ant-hyphen...

Therefore, a `binomial test <https://sites.utexas.edu/sos/guided/inferential/categorical/univariate/binomial/>`_
can determine the p-value of the simulation result in that group, which is the probability
of observing a result equally or less likely, if the simulation's value is correct.
We use two-tailed tests in order to catch errors in either direction.

When simulant-level probabilities of an event vary within a group (for example, if there is a continuous risk factor
of the event), the Bernoulli trials are independent but not identically distributed.
The number of events observed has a `Poisson binomial <https://en.wikipedia.org/wiki/Poisson_binomial_distribution>`_
distribution.
This distribution has the same mean and **lower** variance, relative to a binomial distribution where each trial
has the mean probability.
This simple binomial distribution can be used as an approximation, allowing the use of a binomial test in this situation as well;
due to the variance property, this approximation will increase the specificity of the fuzzy test to be higher than
necessary to achieve the specified family-wise error rate.

> This increase in specificity will cause a decrease in sensitivity.

Reviewer 3: Similarly here, is there a way to show how the use of binomial/poisson would change the specificity/sensitivity visually?

one for each of the values we want to check in the simulation.
In these hypothesis tests, the null hypothesis is that the simulation value matches the V&V target;
rejecting the null hypothesis indicates a problem with the simulation.

Sensitivity and specificity
+++++++++++++++++++++++++++

The **sensitivity** of a check is the probability of it catching
an issue, given that the issue is present.
The **specificity** is the probability of the check passing when
there is no issue present.

In this diagram `from Wikipedia <https://en.wikipedia.org/wiki/Sensitivity_and_specificity>`_, the yellow plane represents the decision boundary: to the left of this boundary, our check
considers the simulation "Healthy"; to the right, our check
considers the simulation "Sick."
The boundary can be chosen arbitrarily, which illustrates the direct tradeoff
between sensitivity and specificity.
By moving it left, we reduce the number of false negatives (missed bugs, in the simulation context), increasing sensitivity.
However, we also increase the number of false positives (false alarms),
decreasing specificity.

.. figure:: PPV,_NPV,_Sensitivity_and_Specificity.svg

>   By Original by Luigi Albert Maria - SVG version of File\:PPV, NPV, Sensitivity and Specificity.pdf, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=99283192

Reviewer 2: ```suggestion
  Original by Luigi Albert Maria - SVG version of File\:PPV, NPV, Sensitivity and Specificity.pdf, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=99283192
```

Author: Yep I did notice this, but since the typo [is on their end](https://commons.wikimedia.org/w/index.php?curid=99283192) (it literally says "Original by Luigi Albert Maria" in the "Author" field) I figured I would keep as-is.
